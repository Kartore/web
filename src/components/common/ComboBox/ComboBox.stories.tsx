import { ComboBox } from '.';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  component: ComboBox,
} satisfies Meta<typeof ComboBox>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
};
