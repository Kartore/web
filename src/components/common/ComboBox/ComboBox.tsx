import type { FC } from 'react';
import { useRef } from 'react';

import { useComboBox, useFilter } from 'react-aria';
import type { ComboBoxStateOptions } from 'react-stately';
import { useComboBoxState } from 'react-stately';

import { Popover } from '~/components/common/Popover';
import { ListBox } from '~/components/common/Select/ListBox';
import { ArrowDropDownIcon } from '~/components/icons';
import { cn } from '~/utils/tailwindUtil';

export type ComboBoxProps = {
	className?: string;
} & ComboBoxStateOptions<object>;

export const ComboBox: FC<ComboBoxProps> = ({ className, ...props }) => {
	const { contains } = useFilter({ sensitivity: 'base' });
	const state = useComboBoxState({
		...props,
		defaultFilter: contains,
	});

	// Setup refs and get props for child elements.
	const buttonRef = useRef(null);
	const inputRef = useRef(null);
	const listBoxRef = useRef(null);
	const popoverRef = useRef(null);

	const { buttonProps, inputProps, listBoxProps, labelProps } = useComboBox(
		{
			...props,
			inputRef,
			buttonRef,
			listBoxRef,
			popoverRef,
		},
		state,
	);

	return (
		<div className={cn('flex items-center justify-between text-sm', className)}>
			<div
				{...labelProps}
				className={cn('font-semibold text-gray-600', labelProps.className)}
			>
				{props.label}
			</div>
			<div
				aria-expanded={state.isOpen}
				className={
					'flex w-1/2 cursor-pointer flex-row rounded border-none bg-gray-100 px-2 py-1 font-semibold text-sm transition-colors hover:bg-gray-200 focus-visible:bg-gray-200 focus-visible:outline-0 aria-expanded:bg-gray-200'
				}
				onClick={() => state.open()}
			>
				<input
					{...inputProps}
					onClick={() => state.open()}
					ref={inputRef}
					className={'w-full flex-1 border-none focus-visible:outline-0'}
				/>
				<button {...buttonProps} ref={buttonRef} className={'cursor-pointer'}>
					<ArrowDropDownIcon aria-hidden={'true'} className={'w-4'} />
				</button>
				{state.isOpen && (
					<Popover
						state={state}
						triggerRef={inputRef}
						popoverRef={popoverRef}
						placement="bottom start"
					>
						<ListBox {...listBoxProps} listBoxRef={listBoxRef} state={state} />
					</Popover>
				)}
			</div>
		</div>
	);
};

ComboBox.displayName = 'ComboBox';
