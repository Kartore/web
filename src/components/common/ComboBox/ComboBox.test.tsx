import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { describe } from 'vitest';

import * as Stories from './ComboBox.stories';

const { Default } = composeStories(Stories);

describe('Component: ComboBox', () => {
	describe('Snapshot', () => {
		it('Default', () => {
			const { asFragment } = render(<Default />);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
