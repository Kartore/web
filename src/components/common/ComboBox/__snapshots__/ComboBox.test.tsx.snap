// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: ComboBox > Snapshot > Default 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-between text-sm"
  >
    <div
      class="font-semibold text-gray-600"
    />
    <div
      aria-expanded="false"
      class="flex w-1/2 cursor-pointer flex-row rounded border-none bg-gray-100 py-1 px-2 text-sm font-semibold transition-colors hover:bg-gray-200 focus-visible:bg-gray-200 focus-visible:outline-0 aria-expanded:bg-gray-200"
    >
      <input
        aria-autocomplete="list"
        aria-expanded="false"
        autocomplete="off"
        autocorrect="off"
        class="w-full flex-1 border-none focus-visible:outline-0"
        id="react-aria-:r2:"
        role="combobox"
        spellcheck="false"
        type="text"
        value=""
      />
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Show suggestions"
        class="cursor-pointer"
        id="react-aria-:r0:"
      >
        <svg
          aria-hidden="true"
          class="w-4"
          viewBox="0 -960 960 960"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M480-360 280-560h400L480-360Z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
